src/init.cpp

```diff
@@ -1248,14 +1248,21 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
         }
     }
 
+    // Check for host lookup allowed before parsing any network related parameters
+    fNameLookup = GetBoolArg("-dns", DEFAULT_NAME_LOOKUP);
+
     bool proxyRandomize = GetBoolArg("-proxyrandomize", DEFAULT_PROXYRANDOMIZE);
     // -proxy sets a proxy for all outgoing network traffic
     // -noproxy (or -proxy=0) as well as the empty string can be used to not set a proxy, this is the default
     std::string proxyArg = GetArg("-proxy", "");
     SetLimited(NET_TOR);
     if (proxyArg != "" && proxyArg != "0") {
-        CService resolved(LookupNumeric(proxyArg.c_str(), 9050));
-        proxyType addrProxy = proxyType(resolved, proxyRandomize);
+        CService proxyAddr;
+        if (!Lookup(proxyArg.c_str(), proxyAddr, 9050, fNameLookup)) {
+            return InitError(strprintf(_("Invalid -proxy address: '%s'"), proxyArg));
+        }
+
+	proxyType addrProxy = proxyType(proxyAddr, proxyRandomize);
```

Comment:
Yes, that'd have to use libevent DNS lookups, which are asynchronous by default, thus someone more complicated to handle. There may be a synchronous way of doing lookups as well, I don't remember if that's exposed outside the http sublibrary.

I'm not sure how important it is in practice, as running torcontrol over a network is extremely discouraged by the Tor developers.

Note that the plan is still to port over the P2P code to libevent, so these interfaces will likely grow together in the future. Might as well wait for that and to it in a similar way.